projects/plugins/jetpack/_inc/lib/class-jetpack-partner-coupon-helper.php

```diff
@@ -0,0 +1,264 @@
+		// Accept and store a partner coupon if present, and redirect to Jetpack connection screen.
+		$partner_coupon = isset( $_GET['jetpack-partner-coupon'] ) ? sanitize_text_field( $_GET['jetpack-partner-coupon'] ) : false; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( $partner_coupon ) {
+			Jetpack_Options::update_options(
+				array(
+					self::$coupon_option => $partner_coupon,
+					self::$added_option  => gmdate( 'Y-m-d H:i:s' ),
+				)
+			);
+
+			if ( Jetpack::connection()->is_connected() ) {
+				wp_safe_redirect( Jetpack::admin_url( 'showCouponRedemption=1' ) );
+			} else {
+				wp_safe_redirect( Jetpack::admin_url() );
+			}
+		}
+	}
+
+	/**
+	 * Purge partner coupon.
+	 *
+	 * We automatically purge partner coupons after a certain amount of time to prevent
+	 * us from unnecessarily promoting a product for months or years in the future.
+	 */
+	public function purge_coupon() {
+		$date = Jetpack_Options::get_option( self::$added_option, '' );
+
+		if ( empty( $date ) ) {
+			return;
+		}
+
+		$date        = strtotime( $date );
+		$purge_after = strtotime( '+1 month' );
+
+		if ( $date > $purge_after ) {
+			Jetpack_Options::delete_option( self::$coupon_option );
+			Jetpack_Options::delete_option( self::$added_option );
+		}
+	}
+
+	/**
+	 * Get partner coupon data.
+	 *
+	 * @return array|bool
+	 */
+	public static function get_coupon() {
+		$coupon_code = Jetpack_Options::get_option( self::$coupon_option, '' );
+
+		if ( ! is_string( $coupon_code ) || empty( $coupon_code ) ) {
+			return false;
+		}
+
+		$partner = self::$instance->get_coupon_partner( $coupon_code );
+
+		if ( ! $partner ) {
+			return false;
+		}
+
+		$preset = self::$instance->get_coupon_preset( $coupon_code );
+
+		if ( ! $preset ) {
+			return false;
+		}
+
+		$product = self::$instance->get_coupon_product( $preset );
+
+		if ( ! $product ) {
+			return false;
+		}
+
+		return array(
+			'coupon_code' => $coupon_code,
+			'partner'     => $partner,
+			'preset'      => $preset,
+			'product'     => $product,
+		);
+	}
+
+	/**
+	 * Get coupon partner.
+	 *
+	 * @param string $coupon_code Coupon code to go through.
+	 * @return array|bool
+	 */
+	private function get_coupon_partner( $coupon_code ) {
+		if ( ! is_string( $coupon_code ) || false === strpos( $coupon_code, '_' ) ) {
+			return false;
+		}
+
+		$prefix             = strtok( $coupon_code, '_' );
+		$supported_partners = $this->get_supported_partners();
+
+		if ( ! isset( $supported_partners[ $prefix ] ) ) {
+			return false;
+		}
+
+		return array(
+			'name'   => $supported_partners[ $prefix ],
+			'prefix' => $prefix,
+		);
+	}
+
+	/**
+	 * Get coupon product.
+	 *
+	 * @param string $coupon_preset The preset we wish to find a product for.
+	 * @return array|bool
+	 */
+	private function get_coupon_product( $coupon_preset ) {
+		if ( ! is_string( $coupon_preset ) ) {
+			return false;
+		}
+
+		$product_details = Jetpack::get_products_for_purchase( true );
```

Comment:
I feel like the "compromise" fits well with the spirit of WordPress. I just moved it to a filter instead.

The most awesome experience here would be an API request to fetch products etc, but because we have a scenario where the user haven't connected their site yet (pre-connection), we're not allowed to make external requests yet (source: Jesse Friedman, based on JITM experience).

```php
// We add the filter in `Jetpack_Redux_State_Helper` before we get the coupon values for `initialState`:
add_filter( 'jetpack_partner_coupon_products', array( self::class, 'get_partner_coupon_product_descriptions' ) );
```

```php
// And then just queue in the products by wrapping "get_products_for_purchase":

public static function get_partner_coupon_product_descriptions() {
	return Jetpack::get_products_for_purchase( true );
}
```